### ISSUE TITLE
feature: abbreviated qualified path to phony targets

### ISSUE THREAD

**Issue by jhwoodyatt** [ocaml-omake/omake#133]
In my day job, we are using Bazel for software construction. (It's a living, shrug.)

One thing Bazel offers that I miss when I'm using OMake is the hierarchical structure of target names, and I've been thinking about what it would take to extend OMake to be more like Bazel in this regard. I think I've got an idea...

OMake today has `.PHONY` targets that are almost _but not quite_ the right thing. To make them more friends, it would be nice if there were a bit of reserved syntax for targets that expanded automatically to hierarchical "phony target" names in the same way that Bazel target names have a hierarchical structure.

I don't have a strong preference for a specific proposal for how that syntax would work, but I could throw one up if it would help:

- Command line target arguments that begin with `^/` would have the `^` replaced with `/.PHONY`. Example: `$ omake ^/foo/bar/baz` at the root directory would be equivalent to `$ (cd foo/bar && omake baz)`.
- Command line target that have a suffix that starts with `:` would have the part of the suffix following the `:` interpreted as a regular expression that matches targets in the `/.PHONY/` namespace corresponding to their relative location. Example: `$ omake ':test-.+'` would invoke all the phony targets in the current subdirectory tree that match the `test-.+` regex.

One could also imagine some logic that provide for handling the combination of both the above the rules in a way that would seem familiar to Bazel users.

**Comment by ANogin** [comment]
Are you aware that  `.PHONY` targets already form a hierarchy that you can use?  Or are you just asking for a uniform syntax that does not require knowing whether the target is .PHONY + regexp capability?

```
% cat OMakeroot
open build/C
.PHONY: do_it
.SUBDIRS: . a b
    do_it:
        echo $(in $(ROOT), $(dir .))
% omake --output-normal -s do_it
a
b
.
% omake --output-normal -s .PHONY/a/do_it
a
% omake --output-normal -s .PHONY/b/do_it
b
% omake --output-normal -s .PHONY/do_it
a
b
.
```

**Comment by jhwoodyatt** [comment]
> Are you aware that .PHONY targets already form a hierarchy that you can use?

Yes! I am, in fact, suggesting it would be nice to have a more compact syntax for referencing `.PHONY` targets. I would use this in my alternative [Conjury](https://bitbucket.org/jhw/conjury/) library so that it would be easier to build a specific subproject from the command line at the root directory of a project.

I suppose it would be fair to think about this feature request in the context of looking at all the innovative features of Bazel (and its cognates, e.g. Buck, Pants, Please, et cetera), and coming up with new features of OMake to make it better compare with those tools.

**Comment by cspiel** [comment]
Having a strong (GNU-) Make background, I see `.PHONY` (and `.FORCE`, `.PRECIOUS`, etc.)
as an attribute of a target, not part of a target's name.  So, I'd prefer **omake's** `.PHONY` being
optional at the command line; from the user's perspective it introduces just a target like any
other.  Aleksey's example then reduces to something like this

```shell
% omake --output-normal -s a/do_it
```

Wrt to the second part of your suggestion, @jhwoodyatt, I'd even extend its reach: what if
*every* target mentioned at the command line *always* undergoes the same expansions?
Every (POSIX-) shell does e.g. "filename expansion" and "brace expansion".  As the targets
of a whatever Make program often are filenames it seems only natural to provide a similar
expansion mechanism to the one implemented in our UN*X shells.

My preferred syntax would be close to shell globs, not regexps, though.

```shell
$ omake 'test/suite-{alpha,production}' # build test/suite-alpha and
                    # test/suite-production
$ omake 'test/suite-alpha/semantics4[0-9][0-9].log' # build (test-)log files in the
                    # "alpha" suite which are in the 400-group of semantics tests
```

**Comment by jhwoodyatt** [comment]
> My preferred syntax would be close to shell globs, not regexps, though.

Sure. That works for me. The target expansion that Bazel provides is really really limited. Wouldn't take much to do better.
